### PR TITLE
Fixed #30160 -- Added support for LZMA and XZ templates to startapp/startproject management commands.

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#30160](https://code.djangoproject.com/ticket/30160).

### CODE REVIEW

File: docs/spelling_wordlist [hunk 1]
```diff
@@ -342,6 +342,7 @@ lowercasing
 lt
 lte
 Luhn
+LZMA
```
Comment (carltongibson) [inline docs/spelling_wordlist]: Looks like this should be lowercase (from the CI failure)

Comment (felixxm) [reply docs/spelling_wordlist]: IMO it should be uppercase in the release notes, since `LZMA` is a shortcut from names "_Lempel-Ziv-Markov algorithm_" :smile: (see https://docs.python.org/3/library/lzma.html).

Comment (carltongibson) [reply docs/spelling_wordlist]: Or that way, yes. 😀

Comment (pope1ni) [reply docs/spelling_wordlist]: Yeah, I started thinking one way, then the other. Started out lowercased because of https://docs.python.org/3/library/tarfile.html

Let me know what we would prefer...

Comment (carltongibson) [reply docs/spelling_wordlist]: Let's go with @felixxm said.

File: docs/releases/3.0.txt [hunk 2]
```diff
@@ -153,6 +153,10 @@ Management Commands
 * The new :option:`compilemessages --ignore` option allows ignoring specific
   directories when searching for ``.po`` files to compile.
 
+* The :option:`startapp --template` and :option:`startproject --template`
+  options now support templates stored in lzma-compressed tarfiles
+  (`*.tar.xz`).
```
Comment (timgraham) [inline docs/releases/3.0.txt]: Use double backticks and update the list of extensions in `docs/ref/django-admin.txt`.

File: django/utils/archive.py [hunk 3]
```diff
@@ -216,11 +216,10 @@ def close(self):
...
-    '.tar': TarArchive,
-    '.tar.bz2': TarArchive,
-    '.tar.gz': TarArchive,
-    '.tgz': TarArchive,
-    '.tz2': TarArchive,
-    '.zip': ZipArchive,
-}
+extension_map = dict.fromkeys((
+    '.tar',
+    '.tar.bz2', '.tbz2', '.tbz', '.tz2',
```
Comment (timgraham) [inline django/utils/archive.py]: Perhaps this refactoring and the addition of additional extensions like tbz2 should be done in a separate commit to highlight that they don't have any tests or documentation.

Comment (pope1ni) [reply django/utils/archive.py]: Ok.

File: docs/releases/3.0.txt (hunk 2 above)
Comment (pope1ni) [reply docs/releases/3.0.txt]: Ah, yes.

File: docs/spelling_wordlist (hunk 1 above)
Comment (pope1ni) [reply docs/spelling_wordlist]: Am referring to this now as an XZ archive -- `.tar.lzma` is considered legacy and `.tar.xz` uses xz-utils which supports the legacy format and the newer LZMA2...

**Comment by pope1ni** [comment]
Hmm. Have updated this, but please don't merge yet. I noticed [`shutil.unpack_archive()`](https://docs.python.org/3/library/shutil.html#shutil.unpack_archive) and want to investigate replacing `django.utils.archive` as it is only used for extracting these templates...

**Comment by pope1ni** [comment]
So I investigated replacing `django.utils.archive.extract()` with `shutil.unpack_archive()`. Unfortunately there are some fixes in the former to resolve two issues:
- If the archive contains a single top-level directory, this is stripped off.
- Python's `zipfile` doesn't correctly apply permissions.

Stripping of the directory is easy. When using `shutil.unpack_archive()` everything works fine for tarfiles w.r.t. permissions, but permissions are wrong when extracting zip files. This is because `shutil._unpack_zipfile()` doesn't handle applying permissions.

For now I'll leave this here for future reference:

```python
def extract(path, to_path):
    shutil.unpack_archive(path, to_path)
 
    # If the files extracted from the archive are contained within a single
    # leading directory, remove that leading directory. This is useful for
    # templates downloaded from GitHub. See #17042 & #22681.
    it = os.scandir(to_path)
    xx = next(it, None)
    if xx and xx.is_dir() and not next(it, None):
        for xy in os.scandir(xx.path):
            shutil.move(xy.path, to_path)
        os.rmdir(xx.path)
 
    # No longer fixes issues with permissions in zip archives.
    # See #26494 & #27628; also https://bugs.python.org/issue15795
    # shutil._unpack_zipfile() is also broken: https://bugs.python.org/issue20907
```

**Comment by pope1ni** [comment]
So I'm going to think about this some more because the permission checks seem inadequate. The file `executable` is only being checked for `u=x` (`S_IOTHX`) but the file is actually `0o775` in the archive. The file `no_permissions` is being checked for `u=r` (`S_IOTHR`) but the file is actually `0o000` in the archive, but is resorting to `0o644` because no permissions are being applied. If the file in the archive were `0o600` -- which ought to be valid because it is "readable" -- it will also end up as `0o644` because we are only checking the permissions for other, not user and group.

A file with no permissions ought to have no permissions. It is only because of issues with `external_attr` being unexpectedly empty in some cases in the fix for [#26494](https://code.djangoproject.com/ticket/26494) that there was a need for [#27628](https://code.djangoproject.com/ticket/27628) which should only have been applied to zip files, not tar files also AFAICT. Having said that, I've just downloaded the supposedly problematic zip from [#27628](https://code.djangoproject.com/ticket/27628) and using `unzip` I don't see any problems...

**Comment by pope1ni** [comment]
I think this is ready for another round of review.

**Comment by carltongibson** [comment]
Ok, thanks Nick. Will put it on the queue for this week coming up.

File: docs/ref/django-admin.txt [hunk 4]
```diff
@@ -1222,7 +1222,7 @@ For example::
 .. django-admin-option:: --template TEMPLATE
 
 Provides the path to a directory with a custom app template file or a path to a
-compressed file (``.tar.gz``, ``.tar.bz2``, ``.tgz``, ``.tbz``, ``.zip``)
```
Comment (felixxm) [inline docs/ref/django-admin.txt]: I don't think that this change is necessary because in the next commit (9ece0f6ec2638d40f1764c9e10e49216687b286f) you added `.tbz` and `.tz2` as aliases of `.tar.bz2`. I would rather remove this commit and add all missing aliases to the `docs/ref/django-admin.txt` as a part of (_" Refs #30160 -- Simplified archive extension map and add other aliases. "_).

File: docs/spelling_wordlist [hunk 5]
```diff
@@ -345,6 +345,7 @@ lowercasing
...
```
Comment (felixxm) [inline docs/spelling_wordlist]: You don't need to add `LZMA` to the wordlist, please remove it.

File: docs/spelling_wordlist [hunk 6]
```diff
@@ -792,4 +793,5 @@ xe
 xgettext
 XSS
 xxxxx
+XZ
```
Comment (felixxm) [inline docs/spelling_wordlist]: You don't need to add `XZ` to the wordlist, please remove it.

**Comment by pope1ni** [comment]
@felixxm Thanks for the guidance. Having the test simplifications first made the addition of support for XZ and LZMA really short and sweet. :guitar:

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
@pope1ni Thanks!
